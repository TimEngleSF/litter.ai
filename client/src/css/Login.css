.login-container {
  width: 100%;
}

.login-logo {
  display: flex;
  width: 100%;
  height: 30dvh;
  background-color: var(--main-green);
  margin-bottom: 1.5rem;
  align-items: center;
  justify-content: center;
}

.login-logo-image{
  max-height: 10rem;
  grid-column-start: 2;
  grid-row-start: 2;
  object-fit: contain;
}

.login-form {

}

.invalid-login {
  color: rgb(255, 0, 0);
  text-align: center;
}

.login-email, .login-password {
  width: 50%;;
  margin: 10px auto;
}

.login-email label, .login-password label {
  color: gray;
  font-size: small;
}

.login-email input, .login-password input {
  display: block;
  width: 100%;
}