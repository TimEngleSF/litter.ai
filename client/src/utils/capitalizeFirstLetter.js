export const capitalizeList = (list) => (
  Object.keys(list).map(
    (key) => [key.charAt(0).toUpperCase() + key.slice(1), list[key]],
  )
);

export const capitalizeWord = (word) => (
  word.charAt(0).toUpperCase() + word.slice(1)
);
