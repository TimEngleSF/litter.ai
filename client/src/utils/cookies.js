import Cookies from 'js-cookie';

export default function loginAuth(loginToken) {
  const token = Cookies.get('authToken');
  if (token) return token;
  Cookies.set('authToken', loginToken, { expires: 7 });
  return token;
}
