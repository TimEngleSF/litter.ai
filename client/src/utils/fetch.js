/* eslint-disable consistent-return */
import Cookies from 'js-cookie';
import URLpath from './URLpath';

export default async function fetchData(path, method, body) {
  const token = Cookies.get('authToken');
  const properties = {
    method,
    headers: {
      'Content-Type': 'application/json',
    },
  };
  // Leaderboard queries for the user's stats if they are logged in
  if (token) { properties.headers.Authorization = `Bearer ${token}`; }
  // Body is the email and password for logging in
  if (body) { properties.body = JSON.stringify(body); }

  try {
    const res = await fetch(URLpath(path), properties);
    const response = await res.json();
    return response;
  } catch (err) {
    console.error(err);
  }
}
