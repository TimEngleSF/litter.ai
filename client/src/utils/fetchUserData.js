/* eslint-disable consistent-return */
import Cookies from 'js-cookie';
import fetchData from './fetch';

export async function fetchLogin(body) {
  try {
    const response = await fetchData('login', 'POST', body);
    if (response.token) {
      Cookies.set('authToken', response.token, { expires: 7 });
      Cookies.set('username', response.user.displayUsername, { expires: 7 });
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLogOut() {
  try {
    await fetchData('logout', 'POST');
    await Cookies.remove('authToken');
    await Cookies.remove('username');
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLeaderboardData(path) {
  try {
    const response = await fetchData(path, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchProfileData(user) {
  try {
    const response = await fetchData(`photo/${user}`, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchRegister(body) {
  try {
    const response = await fetchData('register', 'POST', body);
    if (response.status === 'pending') {
      const loginBody = {
        email: body.email,
        password: body.password,
      };
      const loginResponse = await fetchLogin(loginBody);
      return loginResponse;
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}
