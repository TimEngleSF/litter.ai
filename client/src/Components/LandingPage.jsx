import React from 'react';
import { PiFlowerTulipThin } from 'react-icons/pi';
import { Link } from 'react-router-dom';
import '../css/Landing.css';

export default function LandingPage({ user }) {
  return (
    <>
      <div className="landing-container main-container">
        <div className="landing-top-section">
          <PiFlowerTulipThin className="landing-logo" />
          <div className="landing-text">
            <h1>Welcome to LitterSort</h1>
            <p>With just a photo, sort your garbage and save the world.</p>
          </div>
          <div className="landing-top-buttons lower-buttons">
            <Link to="/capture">
              <button type="button" className="landing-capture-button">
                Capture Picture
              </button>
            </Link>
          </div>
        </div>
        <div className="landing-bottom-section">
          <div className="landing-bottom-buttons lower-buttons">
            <Link to="/leaderboard">
              <button type="button" className="landing-secondary-button">
                Leaderboard
              </button>
            </Link>
            { user
              ? (
                <Link to="/profile">
                  <button type="button" className="landing-secondary-button">
                    Profile
                  </button>
                </Link>
              )
              : (
                <>
                  <Link to="/register">
                    <button type="button" className="landing-secondary-button">
                      Sign Up
                    </button>
                  </Link>
                  <div className="landing-login-text">
                    <p>Already have an account?</p>
                    <Link to="/login">
                      Log in
                    </Link>
                  </div>
                </>
              )}
          </div>
        </div>
      </div>
      <div className="landing-full-bg" />
    </>
  );
}
