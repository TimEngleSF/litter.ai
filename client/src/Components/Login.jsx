import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { FaAngleLeft, FaRegEnvelope } from 'react-icons/fa6';
import { FiLock } from 'react-icons/fi';
import { fetchLogin } from '../utils/fetchUserData';
import '../css/LoginSignUp.css';

export default function Login({ setUser }) {
  const [loginData, setLoginData] = useState({
    email: '',
    password: '',
  });

  const [invalidLogin, setInvalidLogin] = useState(false);
  const navigate = useNavigate();

  const handleLogin = async () => {
    try {
      const response = await fetchLogin(loginData);
      setUser(response.user.displayUsername);
      setInvalidLogin(false);
      navigate('/');
    } catch (err) {
      setInvalidLogin(true);
      console.error(err);
    }
  };

  const handleChange = (e) => {
    e.preventDefault();
    setLoginData({ ...loginData, [e.target.id]: e.target.value });
  };

  const goBack = () => {
    navigate(-1);
  };

  return (
    <div className="login-signup-container main-container">
      <div className="login-signup-header">
        <button className="back-button" type="button" aria-label="Back" onClick={goBack}>
          <FaAngleLeft />
        </button>
        <div className="login-signup-header-text">
          <h3>Sign In</h3>
          <p>Enter your credentials to continue</p>
        </div>
      </div>

      <form className="login-signup-form">
        {invalidLogin
          ? (
            <div className="invalid-login-signup">
              <small>Incorrect username and/or password</small>
            </div>
          )
          : <br className="invalid-login-signup" />}
        <div className="login-signup-email">
          <FaRegEnvelope className="login-signup-icon" />
          <input
            type="email"
            id="email"
            aria-label="Email"
            placeholder="Email"
            value={loginData.email}
            onChange={(e) => handleChange(e)}
          />
        </div>
        <br className="login-signup-form-break" />
        <div className="login-signup-password">
          <FiLock className="login-signup-icon" />
          <input
            type="password"
            id="password"
            aria-label="Password"
            placeholder="Password"
            minLength="4"
            value={loginData.password}
            onChange={(e) => handleChange(e)}
          />
        </div>
      </form>

      <div className="lower-buttons login-signup-buttons">
        <button
          type="button"
          onClick={handleLogin}
          id="login-signup-button"
        >
          Login
        </button>
        <div className="sign-up">
          <small>Don&apos;t have an account?</small>
          <Link
            to="/register"
            className="register-button"
          >
            <small>Sign Up</small>
          </Link>
        </div>
      </div>
    </div>
  );
}
