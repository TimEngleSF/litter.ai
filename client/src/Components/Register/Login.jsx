import React, { useState, useContext } from 'react';
import { Link } from 'react-router-dom';
import UserContext from '../../utils/UserContext';
import { fetchLogin, fetchLogOut } from '../../utils/fetchUserData';

export default function Login({ setUser, user }) {
  const userContext = useContext(UserContext);
  console.log('userContext: ', userContext);
  const [loginData, setLoginData] = useState({
    email: '',
    password: '',
  });

  const handleLogin = async () => {
    try {
      await userContext.login(loginData);
    } catch (err) {
      console.error(err);
    }
    // try {
    //   const userData = await fetchLogin(loginData);
    //   await setUser(userData);
    // } catch (err) {
    //   console.error(err);
    // }
  };

  const handleChange = (e) => {
    e.preventDefault();
    setLoginData({ ...loginData, [e.target.id]: e.target.value });
  };

  const handleCheck = () => {
    console.log('user: ', user);
  };

  return (
    <div>
      <form>
        <label htmlFor="email">
          email:
          <input
            type="text"
            id="email"
            value={loginData.email}
            onChange={(e) => handleChange(e)}
          />
        </label>
        <label htmlFor="password">
          password:
          <input
            type="text"
            id="password"
            value={loginData.password}
            onChange={(e) => handleChange(e)}
          />
        </label>
        <button type="button" onClick={handleLogin}>Login</button>
        <button type="button" onClick={handleCheck}>Test Login Data</button>
        <button type="button" onClick={() => fetchLogOut()}>Log out</button>
        <Link to="/leaderboard"><button type="button">Leaderboard</button></Link>
      </form>
    </div>
  );
}
