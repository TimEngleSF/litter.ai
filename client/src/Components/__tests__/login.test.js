/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  act,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { MemoryRouter as Router, Routes, Route } from 'react-router-dom';
import LandingPage from '../LandingPage';
import Login from '../Login';
import * as fetchUserData from '../../utils/fetchUserData';
// import {
//   mockTotalUploads,
//   mockPlasticUploads,
//   mockMetalUploads,
// } from '../../MockData/mockLeaderboardData';

const mockSetState = jest.fn();

describe('Login component', () => {
  const user = 'Briana30';
  const login = (
    <Router>
      <Login setUser={mockSetState} />
    </Router>
  );

  test('Login page matches the current snapshot', () => {
    const tree = renderer.create(login);
    expect(tree).toMatchSnapshot();
  });

  test('Back button, input fields, login button, and sign up link render to the page', async () => {
    await render(login);
    // TODO: use getByRole and getByTestId for hierarchy
    expect(screen.getByLabelText('Back')).toBeInTheDocument();
    expect(screen.getByLabelText('Email')).toBeInTheDocument();
    expect(screen.getByLabelText('Password')).toBeInTheDocument();
    expect(screen.getByLabelText('Login')).toBeInTheDocument();
    expect(screen.getByLabelText('Sign Up')).toBeInTheDocument();
  });

  // Test invalid login
  test('Invalid login renders an message for the user', () => {
    // expect(screen.getByText('Incorrect username and/or password')).toBeInTheDocument();

    // TODO: Do stuff
    // Displays error in console
  });

  // Test valid login
  test('Valid login sets auth token cookie', () => {
    // TODO: Do stuff
    // Navigates to home page
  });

  // Test back button
  /*
  test('Back button navigates to previous page', () => {
    render(
      <Router>
        <Routes>
          <Route path="/login" element={<Login setUser={mockSetState} />} />
          <Route path="/" element={<LandingPage user={user} />} />
        </Routes>
      </Router>,
    );
    const backBtn = screen.getByTestId('backBtn');
    expect(backBtn).toBeInTheDocument();
  });
    // TODO: Do stuff
  });
  */

  // Test Sign Up button
  test('Sign Up button takes user to registration page', () => {
    // TODO: Do stuff
  });
});
