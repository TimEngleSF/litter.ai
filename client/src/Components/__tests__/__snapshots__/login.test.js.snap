// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Login page matches the current snapshot 1`] = `
<main
  className="login-signup-container main-container"
>
  <section
    className="login-signup-header"
  >
    <button
      aria-label="Back"
      className="back-button"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
        />
      </svg>
    </button>
    <article
      className="login-signup-header-text"
    >
      <h3>
        Sign In
      </h3>
      <p>
        Enter your credentials to continue
      </p>
    </article>
  </section>
  <form
    className="login-signup-form"
    onSubmit={[Function]}
  >
    <br
      className="invalid-login-signup"
    />
    <div
      className="login-signup-email"
    >
      <svg
        className="login-signup-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 112c-8.8 0-16 7.2-16 16v22.1L220.5 291.7c20.7 17 50.4 17 71.1 0L464 150.1V128c0-8.8-7.2-16-16-16H64zM48 212.2V384c0 8.8 7.2 16 16 16H448c8.8 0 16-7.2 16-16V212.2L322 328.8c-38.4 31.5-93.7 31.5-132 0L48 212.2zM0 128C0 92.7 28.7 64 64 64H448c35.3 0 64 28.7 64 64V384c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V128z"
        />
      </svg>
      <input
        aria-label="Email"
        data-testid="emailInput"
        id="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <br
      className="login-signup-form-break"
    />
    <div
      className="login-signup-password"
    >
      <svg
        className="login-signup-icon"
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="11"
          rx="2"
          ry="2"
          width="18"
          x="3"
          y="11"
        />
        <path
          d="M7 11V7a5 5 0 0 1 10 0v4"
        />
      </svg>
      <input
        aria-label="Password"
        id="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <br
      className="login-signup-form-break"
    />
    <section
      className="lower-buttons login-signup-buttons"
    >
      <button
        aria-label="Login"
        className="login-signup-button"
        type="submit"
      >
        Login
      </button>
      <div
        className="sign-up"
      >
        <small>
          Don't have an account?
        </small>
        <a
          aria-label="Sign Up"
          className="register-button"
          href="/register"
          onClick={[Function]}
        >
          <small>
            Sign Up
          </small>
        </a>
      </div>
    </section>
  </form>
</main>
`;
